### PR TITLE
feat: Linting in entity properties and methods

### LINKED ISSUE #10695: [Bug]: The linter is not showing error warnings for invalid properties
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Please see this image. Here I wrote `selectedItem` as `selectedItems` by mistake but the linter did not show any error message.  This made debugging the error very difficult.
image

### Steps To Reproduce

1. Create a list as List1
2. Create a JS Object
3. Access `selectedItem` property on of `List1` as `List1.selectedItems`
4. The linter should have thrown the error

### Environment

Production

### Version

Cloud

### LINKED ISSUE #11663: [Bug]: Invalid JS function names are not validated for errors
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Function names that don't exist when bound to a widget (using mustache binding) don't get validated for error, i.e, linting does not show up.

### Steps To Reproduce

1. Create a new JS object
2. Bind it's function `JSObject1.myFun1()` to a widget
3. Now change it to `JSObject1.myFun11()` - observe that linting does not show up and `Evaluate value` does not display any value
4. Bind JSObject1.myFun2() to a widget and delete this binding and observe that no lint error is displayed. Only execution on action trigger does an error show up

### Environment

Production

### Version

Cloud

### LINKED ISSUE #14895: [Bug]-[28000]:Binding shows successful even for invalid JS
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

Writing invalid code shows a successful state all across the editor

### Steps To Reproduce

1. Drag a table widget
2. Drag a Select widget
3. Bind {{Table1.test}} in the options field
4. See that it shows success
5. Add a JSObject and add the same invalid code in a function and notice it does not show an error

### Public Sample App

_No response_

### Version

Cloud

### LINKED ISSUE #15407: [Bug]: Linting is not reactive when path is incorrect
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

See loom recording [here](https://www.loom.com/share/610c4432407548d29ba19107eb0192fc)

### Steps To Reproduce

1 Drag a button widget
2. Create an entity (Api, Query or JSObject)
3. In any field, add an incorrect path eg. `entityname.tetete`
4. Delete entity, and see that linting doesn't get updated

### Public Sample App

_No response_

### Version

Cloud

### LINKED ISSUE #16384: [Task]: Update extraction logic of AST to support variable extraction as well
### Is there an existing issue for this?

- [X] I have searched the existing issues

### SubTasks

- Update Extraction function `extractIdentifiersFromCode` in AST module to return identifiers with variables as well. 
- Move Jest Test case to AST shared module
- Remove redundant AST code from `client`

### LINKED ISSUE #16427: [Bug]: AST parsing returns local variable if child property is accessed in script
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

`extractIdentifiersFromCode` function in AST shared module if called with params `script` as `(function(par){ let Api3 = {}; return [Api1.data, Api2.data, Api3.data, par.a]; })()` and `evaluationVersion`  as  `2`, 

Current Returned value:

`{"identifiers":["Api1.data","Api2.data","Api3.data","par.a"],"variables":["Api3"]}`  

Expected Returned value:

`{"identifiers":["Api1.data","Api2.data"],"variables":["Api3"]}`

### Steps To Reproduce

Manually or starting RTS server and sending a request `/rts-api/v1/ast/single-script-identifiers` with params

### Public Sample App

_No response_

### Version

Self Hosted

### PR DESCRIPTION
**Pull request by ohansFavour**
### Description

This PR introduces linting of entity's properties. Now, the linter picks up errors for invalid properties of an entity.

180985306-1748a218-6b79-49bf-b3da-f1ff7765915e

### Dependency map update timings

CleanShot 2022-09-03 at 19 55 21 2@2x

Major addition

 A new object (`unusedIdentifiersList`), that stores the list of unreferenced identifiers (unused identifiers) throughout the application. This is needed to prevent complete traversal of the dataTree when an entity or a new path is added to the dataTree while searching for references in old bindings.

Improvements.

1.   Performance

     - [x]  Prevent traversal of dataTree for every addition of entity/path while searching for references in old bindings.
     - [x]  Prevent extraction of references for all paths in the dependency map for every change.
     
###
 2. - [x] Linting in trigger fields only depends on top-level entities
 
 3. - [x] Javascript built-in classes should not be valid entity names. Eg, naming a JSObject as "Promise", makes Promise.all() unusable within it.

 4. - [x] Support linting of invalid properties

 5.   Incorrect extraction of identifiers from code  
       

         - [x] Remove all identifiers which are children of global dependencies like moment, lodash, xmlParser, forge.
         - [x] Remove all identifiers which are children of Appsmith functions like showAlert, setInterval.
         - [x] Remove all identifiers which are children of browser Apis like console.log, Object.keys, 
         - [x] Remove all member expressions which are derived from params.

         
        Eg. In code {{Api1.data.map((arg) => arg.data)}} ,  `arg.data` should not be a valid identifier.
   
        Edge case    
        - [ ] Only remove paths that are "actual" children of params
   
```
{{
function(){
 const test1  =   ["name", "school"].map((Api1)=>{
    return Api1.trim() // Api1.trim should not be a valid identifier since it's a child of param Api1
       })


 const test2 = ["name", "school"].map((detail)=>{
    return Api1.name + detail  // Api1.name should be a valid identifier since "Api1" here is not within the scope of that declared in test1
    })
   
   }()
}}

```
  
- [x] Remove all identifiers/member expressions which are derived from declared variables

Eg.

```
{{
function(){
 const test1  =   ["name", "school"].map((a)=>{
const Api1 = a.trim();
    return Api1.split("") // Api1.split should not be a valid identifier since it's a child of declared variable Api1
       })


 const test2 = ["name", "school"].map((b)=>{
    return test1.join()  // test1.join should not be a valid identifier
    })
   
   }()
}}
```
Edge case

- [ ]  Only remove paths that are "actual" children of declared variables.

```
{{
function(){
 const test1  =   ["name", "school"].map((a)=>{
const Declared1 = a
return Declared1 
       })


const test2 = ["name", "school"].map((b)=>{
    return Declared1.name  // Declared1.name should be a valid identifier since it's not within the scope of "Declared1" in test1 and could possibly be an entity's name.
    })
   
   }()
}}
```  

- Update Extraction function `extractIdentifiersFromCode` to `extractInfoFromCode` in AST module to return identifiers with variables as well.
- Move Jest Test case to AST shared module
- Removed redundant AST code from client
 
 
Fixes #11663
Fixes #10695
Fixes #14895
Fixes #15407 
Fixes #16384
Fixes #16427

## Type of change

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
-**Test Plan**-
- [ ] https://github.com/appsmithorg/TestSmith/issues/2060

WIP

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by Rishabh-Rathod — COMMENTED** [review]
initial review

File: app/client/src/constants/WidgetValidation.ts [hunk 1]
```diff
@@ -172,3 +172,41 @@ export const WINDOW_OBJECT_METHODS = {
   setTimeout: "setTimeout",
   stop: "stop",
 };
+
+export const INTERNAL_IDENTIFIERS = ["appsmith"];
+
+export const GLOBAL_SCOPE_OBJECTS = {
```
[the remark this answers is not in the document]
Comment (jnikhila) [reply app/client/src/constants/WidgetValidation.ts:178]: @Rishabh-Rathod - The [doc here talks about the entity names](https://github.com/appsmithorg/appsmith/pull/16171#pullrequestreview-1094507581). Do you think we can improvise it to capture details, or should we add other error types to give a holistic view?

File: app/client/src/workers/DependencyMap/index.ts [hunk 2]
```diff
@@ -21,22 +22,34 @@ import {
...
+
+interface CreateDependencyMap {
+  dependencyMap: DependencyMap;
+  triggerFieldDependencyMap: DependencyMap;
+  /** Keeps track of all unused Identifiers present in bindings throughout the page.
+   *  We keep this list so that we don't have to traverse the entire dataTree when
+   *  a new entity or path is added to the datatree in order to determine if an old binding has become valid
+   *  because an entity or path is newly added.
+   * */
+  unusedIdentifiers: DependencyMap;
```
Comment (Rishabh-Rathod) [inline app/client/src/workers/DependencyMap/index.ts:43]: Suggestions for code refactor, we could skip it for this PR if seems like lot of changes.
  
Shall we move all dependencyMap state to this file be creating DependencyMap class with
state: `evalDependencyMap` , `triggerFieldDependencyMap` , `unusedIdentifiersMap` 
method: `create`, `update`, getters for each state

**Review by Rishabh-Rathod — COMMENTED** [review]
review phase 2/n

File: app/shared/ast/src/constants.ts [hunk 3]
```diff
@@ -4,27 +4,194 @@ export const ECMA_VERSION = 11;
...
+  screen: 'screen',
+  screenLeft: 'screenLeft',
+  screenTop: 'screenTop',
+  screenY: 'screenY',
+  scrollX: 'scrollX',
+  scrollY: 'scrollY',
+  self: 'self',
+  status: 'status',
+  top: 'top',
+};
```
Comment (Rishabh-Rathod) [inline app/shared/ast/src/constants.ts:132]:
```suggestion
export const WINDOW_OBJECT_PROPERTIES = {
  console: 'console',
  localStorage: 'localStorage',
  name: 'name',
};
```
As the evaluations in happening in worker, we can avoid DOM APIs

**Review by Rishabh-Rathod — COMMENTED** [review]
review phase 3/n

File: app/shared/ast/index.test.ts [hunk 4]
```diff
@@ -220,11 +220,59 @@ describe("getAllIdentifiers", () => {
...
+          const firstApiRun = Api1.run();
+          const secondApiRun = Api2.run();
+          const randomNumber = Math.random();
+          return Promise.all([firstApiRun, secondApiRun])
+        }()`,
+        expectedResults: ["Api1.run", "Api2.run"],
+      },
+      // Global dependencies should not be valid identifiers
+      {
+        script: `function(){
```
Comment (AmanAgarwal041) [inline app/shared/ast/index.test.ts:235]: If you run test cases locally, this test case would fail. Please check this test case

File: app/client/src/workers/DependencyMap/index.ts [hunk 5]
```diff
@@ -54,39 +67,62 @@ export function createDependencyMap(
...
-          message: (error as Error).message,
-          context: {
-            script: path,
-          },
-        });
-        return [];
-      }
+    const {
+      errors,
+      references,
```
[the remark this answers is not in the document]
Comment (ohansFavour) [reply app/client/src/workers/DependencyMap/index.ts]: As discussed, the new nomenclature

references => All references to entities in a binding

validReference => references that that are currently valid.

invalidReferences =>  references that that are currently invalid

invalidReferencesMap => Map of nodes to all currently invalid references in their bindings.

File: app/client/src/workers/DependencyMap/index.ts [hunk 6]
```diff
@@ -234,41 +402,98 @@ export const updateDependencyMap = ({
...
-                dataTreeEvalRef.triggerFieldDependencyMap[
-                  triggerField
-                ] = dataTreeEvalRef.triggerFieldDependencyMap[
-                  triggerField
-                ].filter((field) => field !== entityName);
-              });
-            }
+          Object.keys(dataTreeEvalRef.triggerFieldDependencyMap).forEach(
+            (dependencyPath) => {
+              didUpdateDependencyMap = true;
```
[the remark this answers is not in the document]
Comment (ohansFavour) [reply app/client/src/workers/DependencyMap/index.ts]: That was a mistake

File: app/client/src/workers/DependencyMap/index.ts (hunk 2 above)
Comment (ohansFavour) [reply app/client/src/workers/DependencyMap/index.ts]: We will have to discuss this with the team, after which we can proceed. Let's do this in a different PR (after discussing with the team).

File: app/shared/ast/src/constants.ts [hunk 7]
```diff
@@ -4,27 +4,194 @@ export const ECMA_VERSION = 11;
...
+  requestAnimationFrame: 'requestAnimationFrame',
+  resizeBy: 'resizeBy',
+  resizeTo: 'resizeTo',
+  scroll: 'scroll',
+  scrollBy: 'scrollBy',
+  scrollTo: 'scrollBy',
+  setInterval: 'setInterval',
+  setTimeout: 'setTimeout',
+  stop: 'stop',
+};
```
[the remark this answers is not in the document]
Comment (ohansFavour) [reply app/shared/ast/src/constants.ts]: Nice point, I have modified this.

File: app/client/src/workers/Lint/utils.ts [hunk 8]
```diff
@@ -49,9 +56,8 @@ export const pathRequiresLinting = (
     (isAction(entity) || isWidget(entity) || isJSAction(entity)) &&
     isPathADynamicBinding(entity, propertyPath);
   const requiresLinting =
-    isADynamicBindingPath &&
-    (isDynamicValue(unEvalPropertyValue) ||
-      (isJSAction(entity) && propertyPath === "body"));
+    (isADynamicBindingPath && isDynamicValue(unEvalPropertyValue)) ||
+    isJSAction(entity);
```
Comment (Rishabh-Rathod) [inline app/client/src/workers/Lint/utils.ts:60]: We do want to only line JSObject.body correct?`&& propertyPath === "body")` 
Can you help in understanding why was this removed?

File: app/shared/ast/package.json [hunk 9]
```diff
@@ -37,7 +40,9 @@
     "rollup-plugin-generate-package-json": "^3.2.0",
     "rollup-plugin-peer-deps-external": "^2.2.4",
     "rollup-plugin-typescript2": "^0.32.0",
-    "ts-jest": "27.0.0"
+    "ts-jest": "27.0.0",
+    "@types/moment-timezone": "^0.5.10",
+    "@types/node-forge": "^0.10.0"
```
Comment (Rishabh-Rathod) [inline app/shared/ast/package.json:45]: We could undo above changes if removing extraLibraries.

File: app/client/src/workers/Lint/utils.ts [hunk 10]
```diff
@@ -201,3 +216,49 @@ const getValidLintErrors = (lintErrors: LintError[], scriptPos: Position) => {
...
+
+  const invalidPropertyErrors = invalidTopLevelMemberExpressions.map(
+    ({ object, property }) => {
+      const propertyName = isLiteralNode(property)
+        ? property.value
+        : property.name;
+      const objectStartLine = object.loc.start.line - 1;
+      const propertyStartColumn = !isLiteralNode(property)
+        ? property.loc.start.column + 1
+        : property.loc.start.column + 2;
```
Comment (Rishabh-Rathod) [inline app/client/src/workers/Lint/utils.ts:243]: Please add comment to explain the logic

File: app/client/src/workers/Lint/index.ts [hunk 11]
```diff
@@ -54,10 +54,9 @@ export const lintTree = (args: LintTreeArgs) => {
     // We are only interested in paths that require linting
     if (!pathRequiresLinting(unEvalTree, entity, fullPropertyPath)) return;
     if (isATriggerPath(entity, propertyPath))
-      return triggerPaths.push(fullPropertyPath);
+      return triggerPaths.add(fullPropertyPath);
     if (isJSAction(entity))
-      return bindingPathsRequiringFunctions.push(fullPropertyPath);
-
+      return bindingPathsRequiringFunctions.add(`${entityName}.body`);
```
Comment (ohansFavour) [inline app/client/src/workers/Lint/index.ts:59]: @Rishabh-Rathod , here, the right logic has been implemented. For example, when JSObject.myFun1 is to be linted, we replace that with JSObject.body. Previous implementation was totally overlooking functions and variables in the JSObject.

File: app/client/src/workers/Lint/utils.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (ohansFavour) [reply app/client/src/workers/Lint/utils.ts:60]: https://github.com/appsmithorg/appsmith/pull/16171/files#r965014236

File: app/shared/ast/src/constants/global.ts [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+    // We are removing some functionalities of node-forge because they wont
+    // work in the worker thread
+    lib: _.omit(forge, ['tls', 'http', 'xhr', 'socket', 'task']),
+    version: '1.3.0',
+    docsURL: 'https://github.com/digitalbazaar/forge',
+    displayName: 'forge',
+  },
+];
+
+export const GLOBAL_FUNCTIONS = {
```
[the remark this answers is not in the document]
Comment (ohansFavour) [reply app/shared/ast/src/constants/global.ts:174]: I also share the same sentiment, but the problem here is that the server also uses the extractInfoFromCode. This might require us to maintain the list of global function accessors both on the server and the client.

[the remark this answers is not in the document]
Comment (Rishabh-Rathod) [reply app/shared/ast/src/constants/global.ts:174]: SCR-20220907-ua6

I wasn't able to find where `extractInfoFromCode` is used in `shared/ast` code. 

Also, we could have a configuration API where we pass such a function and the `shared/ast` code can then consume it.

[the remark this answers is not in the document]
Comment (ohansFavour) [reply app/shared/ast/src/constants/global.ts:174]: Here (https://github.com/appsmithorg/appsmith/pull/16252). Apis will be created using it.

File: app/client/src/workers/Lint/index.ts (hunk 11 above)
Comment (ohansFavour) [reply app/client/src/workers/Lint/index.ts:59]: Will add a comment, as you can see, what is being added to the bindingPathsRequiringFunctions is  `${entityName}.body`

Comment (Rishabh-Rathod) [reply app/client/src/workers/Lint/index.ts:59]: Sorry, my bad. I was checking in other parts of the code. ( not sure why )

**Comment by ohansFavour** [comment]
Only flaky tests are failing now. @Rishabh-Rathod you can re-review.

File: app/shared/ast/package.json (hunk 9 above)
Comment (Rishabh-Rathod) [reply app/shared/ast/package.json:42]: Reminder comment.

File: yarn.lock [hunk 13]
```diff
@@ -0,0 +1,4 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
+# yarn lockfile v1
+
+
```
Comment (Rishabh-Rathod) [inline yarn.lock:4]: Is this intentional?

**Review by Rishabh-Rathod — DISMISSED** [review]
LGTM.

Please check the below comments.

File: yarn.lock (hunk 13 above)
Comment (ohansFavour) [reply yarn.lock:4]: Yes

**Review by Rishabh-Rathod — APPROVED** [review]

**Review by AmanAgarwal041 — DISMISSED** [review]

**Review by Rishabh-Rathod — APPROVED** [review]
